src/vidpf.rs

```diff
@@ -0,0 +1,676 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Verifiable Incremental Distributed Point Function (VIDPF).
+//!
+//! The VIDPF construction is specified in [[draft-mouris-cfrg-mastic]] and builds
+//! on techniques from [[MST23]] and [[CP22]] to lift an IDPF to a VIDPF.
+//!
+//! [CP22]: https://eprint.iacr.org/2021/580
+//! [MST23]: https://eprint.iacr.org/2023/080
+//! [draft-mouris-cfrg-mastic]: https://datatracker.ietf.org/doc/draft-mouris-cfrg-mastic/01/
+
+use core::{
+    borrow::Borrow,
+    iter::zip,
+    marker::PhantomData,
+    ops::{Add, AddAssign, Sub},
+};
+
+use rand_core::RngCore;
+use std::io::Cursor;
+use subtle::{Choice, ConditionallyNegatable, ConditionallySelectable};
+
+use crate::{
+    codec::{CodecError, Decode, Encode},
+    field::FieldElement,
+    idpf::{conditional_select_seed, conditional_xor_seeds, xor_seeds, IdpfInput, IdpfValue},
+    vdaf::xof::{Seed as XofSeed, Xof, XofFixedKeyAes128, XofTurboShake128},
+};
+
+/// Represents the domain of a point function.
+pub type VidpfInput = IdpfInput;
+
+impl VidpfInput {
+    fn iter_controlbit(&self) -> impl DoubleEndedIterator<Item = ControlBit> + '_ {
+        self.iter().rev().map(|a| ControlBit::from(a as u8))
+    }
+}
+
+/// Represents the codomain of a point function.
+pub trait VidpfValue: IdpfValue + Eq
+where
+    for<'a> &'a Self: Add<Output = Self> + Sub<Output = Self>,
+{
+}
+
+/// Creates a VIDPF instance.
+#[must_use]
+pub const fn new<V>(nonce: Vec<u8>) -> Vidpf<V>
+where
+    V: VidpfValue,
+    for<'a> &'a V: Add<Output = V> + Sub<Output = V>,
+{
+    Vidpf {
+        nonce,
+        phantom: PhantomData,
+    }
+}
+
+/// Contains the parameters of a VIDPF instance.
+pub struct Vidpf<V>
+where
+    // Represents the codomain of a point function.
+    V: VidpfValue,
+    for<'a> &'a V: Add<Output = V> + Sub<Output = V>,
+{
+    // Used to cryptographically bind some information.
+    nonce: Vec<u8>,
+
+    phantom: PhantomData<V>,
+}
+
+impl<V> Vidpf<V>
+where
+    V: VidpfValue,
+    for<'a> &'a V: Add<Output = V> + Sub<Output = V>,
```

Comment:
The IDPF module avoids all of this by passing around owned field elements, etc., and using Add on that. When it uses a borrowed value out of the correction words, it does so using `conditional_select()`, which copies from a reference to an owned value as part of its operation.

Some additions could be rewritten in a similar way; to the extent that VIDPF requires using values of this type more than once, it would be possible to clone first, and then use Add on owned values. I tried replacing the HRTB in the where clause with `Clone + for<'a> Add<&'a Self, Output = Self> bounds on `VidpfValue` itself, and there were three callsites I had to tweak. The calculation of `w_cw` can be rearranged so that `w_i_1` is on the left hand side, and that is an owned value coming out of `convert()`. `beta` can stay as a borrowed value on the right hand side. The calculation of `y_i` could be changed to conditionally select between zero and the correction word, and then add `w_i` to that. This is how the IDPF module does the calculation, and it takes ownership of w_i, using it just once, and copies out of the w_cw borrow. Lastly, `verify()` probably just needs a clone on the left hand argument.

Reply:
thanks for the tip